mantidproject/mantid
# Project Recovery - Offer to recover if Mantid exited incorrectly

**Description of work.**
This is the second major PR for project recovery. It adds the following features:
- Remove checkpoints on graceful exit
- Only save, if there is something to save
- Offer to recover a script if a checkpoint folder is found
- Converted `OrderWorkspaceHistory` from a script to a Mantid Alg
- Add option to ScriptBuilder to append timestamps
- Add function to get Mantid start time
- We only save out commands since Mantid start

**Still pending**
- Bug where a race condition occurs if a workspace is renamed or removed whilst project recovery is executing
- Some edge cases when project recovery has completed (do we restart thread)
- Handling the configuration value less than 0
- Opening GUI windows which were previously closed

**To test:**
- Open Mantid generate some workspaces
- Group a  workspace together, to check that duplicate elimination works
- Exit mantid forcefully (kill `PID` for example)
- Check that loading back in correctly asks about project recovery
- Check that both the `Open in script editor` and `No` option clear the recovery checkpoints after they complete
- Check that recovery starts saving again if `No` is selected

------
- Add an error to the `OrderWorkspaceHistory` alg, for example `str("hello") + int (1)`
- Attempt project recovery
- Ensure that the checkpoints are not deleted if it fails
-----------
- Create some sample workspaces and perform some methods on them
- Save the project normally
- Open / Close Mantid
- Load the project normally
- Perform some more commands
- Kill Mantid
- Reload and recover
- Check that only commands in the second instance were saved

Does this update require release notes?

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
